sdk/core/azure-core/src/main/java/com/azure/core/util/IOUtils.java

```diff
@@ -0,0 +1,118 @@
+        }
+        return new AsynchronousFileChannelAdapter(fileChannel, position);
+    }
+
+    /**
+     * Transfers bytes from {@link ReadableByteChannel} to {@link WritableByteChannel}.
+     * @param source A source {@link ReadableByteChannel}.
+     * @param destination A destination {@link WritableByteChannel}.
+     * @throws IOException When I/O operation fails.
+     * @throws NullPointerException When {@code source} is null.
+     * @throws NullPointerException When {@code destination} is null.
+     */
+    public static void transfer(ReadableByteChannel source, WritableByteChannel destination) throws IOException {
+        Objects.requireNonNull(source, "'source' must not be null");
+        Objects.requireNonNull(source, "'destination' must not be null");
+        ByteBuffer buffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE);
+        int read;
+        do {
+            buffer.clear();
+            read = source.read(buffer);
+            buffer.flip();
+            while (buffer.hasRemaining()) {
+                destination.write(buffer);
+            }
+        } while (read >= 0);
+    }
+
+    /**
+     * Transfers bytes from {@link ReadableByteChannel} to {@link AsynchronousByteChannel}.
+     * @param source A source {@link ReadableByteChannel}.
+     * @param destination A destination {@link AsynchronousByteChannel}.
+     * @return A {@link Mono} that completes when transfer is finished.
+     * @throws NullPointerException When {@code source} is null.
+     * @throws NullPointerException When {@code destination} is null.
+     */
+    public static Mono<Void> transferAsync(ReadableByteChannel source, AsynchronousByteChannel destination) {
+        Objects.requireNonNull(source, "'source' must not be null");
+        Objects.requireNonNull(source, "'destination' must not be null");
+        return Mono.create(sink -> sink.onRequest(value -> {
+            ByteBuffer buffer = ByteBuffer.allocate(DEFAULT_BUFFER_SIZE);
+            try {
+                transferAsynchronously(source, destination, buffer, sink);
+            } catch (IOException e) {
+                sink.error(e);
+            }
+        }));
+    }
+
+    private static void transferAsynchronously(
+        ReadableByteChannel source, AsynchronousByteChannel destination,
+        ByteBuffer buffer, MonoSink<Void> sink) throws IOException {
+        buffer.clear();
+        int read = source.read(buffer);
+        if (read >= 0) {
+            buffer.flip();
+            destination.write(buffer, buffer, new CompletionHandler<Integer, ByteBuffer>() {
+                @Override
+                public void completed(Integer result, ByteBuffer attachment) {
+                    try {
+                        // This is not a classic recursion.
+                        // I.e. it happens in completion handler not on call stack.
+                        transferAsynchronously(source, destination, buffer, sink);
```

Comment:
Doesn't this need to handle the case where the full buffer isn't written?

Reply:
Good catch. I'll fix and add test.